sbin/slattach/slattach.8
.\"	$NetBSD: slattach.8,v 1.22 1999/09/12 18:47:11 kleink Exp $
.\"
.\" Copyright (c) 1986, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)slattach.8	8.2 (Berkeley) 4/1/94
.\"
.Dd September 3, 1999
.Dt SLATTACH 8
.Os BSD 4.3
.Sh NAME
.Nm slattach ,
.Nm slip
.Nd attach serial lines as network interfaces
.Sh SYNOPSIS
.Nm
.Op Fl t Ar ldisc
.Op Fl hHlmn
.Op Fl s Ar baudrate
.Ar ttyname
.Sh DESCRIPTION
.Nm
is used to assign a tty line to a network interface,
and to define the network source and destination addresses.
The following operands are supported by
.Nm "" :
.Bl -tag -width Ar
.It Fl h
Turn on RTS/CTS flow control.  By default, no flow control is done.
.It Fl H
Turn on DTR/CTS flow control.  By default, no flow control is done.
.It Fl l
Turn on the CLOCAL flag, making it possible to run SLIP on a cable
without modem control signals (e.g. DTR, DSR, DCD).
.It Fl m
Maintain modem control signals after closing the line.  Specifically,
this disables HUPCL.
.It Fl n
Don't detach from invoking tty.
.It Fl s Ar baudrate
Specifies the speed of the connection.  If not specified, the
default of 9600 is used.
.It Fl t Ar ldisc
Specifies the line discipline to use for the tty.  Supported line
disciplines are
.Dq slip
(creates a
.Xr sl 4
instance) and
.Dq strip
(creates a
.Xr strip 4
instance).
If this option is not specified, the default is
.Dq slip .
.It Ar ttyname
Specifies the name of the tty device.
.Ar Ttyname
should be a string of the form
.Ql ttyXX ,
or
.Ql /dev/ttyXX .
.El
.Pp
In addition, the following flags to
.Xr ifconfig 8
control various properties of a SLIP link:
.Bl -tag -width Ar
.It Ar link0
Turn on Van Jacobsen header compression.
.It Ar -link0
Turn off header compression.
.It Ar link1
Don't pass through ICMP packets.
.It Ar -link1
Do pass through ICMP packets.
.It Ar link2
If a packet with a compressed header is received, automatically enable
compression of outgoing packets.
.It Ar -link2
Don't auto-enable compression.
.El
.Pp
Only the super-user may attach a network interface.
.Pp
To detach the interface, use
.Dq Li ifconfig interface-name down
after killing off the
.Nm
process.
.Ar Interface-name
is the name that is shown by
.Xr netstat 1 .
.Sh EXAMPLES
.Bd -literal -offset indent -compact
slattach ttyh8
slattach \-s 4800 /dev/tty01
.Ed
.Sh DIAGNOSTICS
Messages indicating the specified interface does not exist, the
requested address is unknown, the user is not privileged and
tried to alter an interface's configuration.
.Sh SEE ALSO
.Xr netstat 1 ,
.Xr daemon 3 ,
.Xr netintro 4 ,
.Xr sl 4 ,
.Xr strip 4 ,
.Xr ifconfig 8 ,
.Xr rc 8
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.3 .


sbin/slattach/slattach.c
/*	$NetBSD: slattach.c,v 1.22 1999/09/03 13:31:29 proff Exp $	*/

/*
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Rick Adams.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1988, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)slattach.c	8.2 (Berkeley) 1/7/94";
#else
__RCSID("$NetBSD: slattach.c,v 1.22 1999/09/03 13:31:29 proff Exp $");
#endif
#endif /* not lint */

#include <sys/param.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/ioctl.h>

#include <net/if.h>
#include <netinet/in.h>

#include <err.h>
#include <fcntl.h>
#include <netdb.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>

int	speed = 9600;
int	slipdisc = SLIPDISC;

char	devicename[32];

int	main __P((int, char *[]));
int	ttydisc __P((char *));
void	usage __P((void));


int
main(argc, argv)
	int argc;
	char *argv[];
{
	int fd;
	char *dev = argv[1];
	struct termios tty;
	tcflag_t cflag = HUPCL;
	int ch;
	sigset_t sigset;
	int opt_detach = 1;

	while ((ch = getopt(argc, argv, "hHlmns:t:")) != -1) {
		switch (ch) {
		case 'h':
			cflag |= CRTSCTS;
			break;
		case 'H':
			cflag |= CDTRCTS;
			break;
		case 'l':
			cflag |= CLOCAL;
			break;
		case 'm':
			cflag &= ~HUPCL;
			break;
		case 'n':
			opt_detach = 0;
			break;
		case 's':
			speed = atoi(optarg);
			break;
		case 'r': case 't':
			slipdisc = ttydisc(optarg);
			break;
		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 1)
		usage();

	dev = *argv;
	if (strncmp(_PATH_DEV, dev, sizeof(_PATH_DEV) - 1)) {
		(void)snprintf(devicename, sizeof(devicename),
		    "%s%s", _PATH_DEV, dev);
		dev = devicename;
	}
	if ((fd = open(dev, O_RDWR | O_NDELAY)) < 0)
		err(1, "%s", dev);
	tty.c_cflag = CREAD | CS8 | cflag;
	tty.c_iflag = 0;
	tty.c_lflag = 0;
	tty.c_oflag = 0;
	tty.c_cc[VMIN] = 1;
	tty.c_cc[VTIME] = 0;
	cfsetspeed(&tty, speed);
	if (tcsetattr(fd, TCSADRAIN, &tty) < 0)
		err(1, "tcsetattr");
	if (ioctl(fd, TIOCSDTR, 0) < 0)
		warn("TIOCSDTR");
	if (ioctl(fd, TIOCSETD, &slipdisc) < 0)
		err(1, "TIOCSETD");
	if (opt_detach && daemon(0, 0) != 0)
		err(1, "couldn't detach");
	sigemptyset(&sigset);
	for (;;)
		sigsuspend(&sigset);
}

int
ttydisc(name)
     char *name;
{
	if (strcmp(name, "slip") == 0)
		return(SLIPDISC);
#ifdef STRIPDISC
	else if (strcmp(name, "strip") == 0)
  		return(STRIPDISC);
#endif
	else
		usage();
	/* NOTREACHED */
	return -1;
}

void
usage()
{
	extern char *__progname;

	(void)fprintf(stderr,
	    "usage: %s [-t ldisc] [-hHlm] [-s baudrate] ttyname\n",
		__progname);
	exit(1);
}
